### PR TITLE
Convert to specs v0.9

### PR DESCRIPTION
**Pull request by omni-viral**
Convert to specs v0.9.1

### CODE REVIEW

File: src/ecs/resources/broadcaster.rs [hunk 1]
```diff
@@ -16,17 +16,15 @@
 //!     type Storage = VecStorage<UserComponent>;
 //! }
 //!
-//! fn main() {
-//!     use amethyst::ecs::Gate;
-//!
+//! fn main() {//!
```
Comment (torkleyy) [inline src/ecs/resources/broadcaster.rs]: I don't think the `//!` is necessary here.

File: src/ecs/resources/broadcaster.rs [hunk 2]
```diff
@@ -35,11 +33,7 @@
 //! }
 //! ```
 
-use std::sync::RwLockReadGuard;
-
-use ticketed_lock::ReadTicket;
-
-use ecs::{World, Component, EntityBuilder, GatedStorage, Allocator, MaskedStorage, Join};
+use ecs::{World, Component, EntityBuilder, Join, ReadStorage};
```
Comment (torkleyy) [inline src/ecs/resources/broadcaster.rs]: We didn't have it before, but it would be nice to sort these imports, too.

File: src/ecs/resources/broadcaster.rs [hunk 3]
```diff
@@ -63,20 +57,23 @@ impl Broadcaster {
...
     pub fn clean(&mut self) {
-        for entity in self.world.entities().join() {
-            self.world.delete_later(entity);
+        let entities = {
+            self.world.entities().join().collect::<Vec<_>>()
+        };
+        for entity in entities {
+            self.world.delete_entity(entity);
         }
         self.world.maintain();
```
Comment (torkleyy) [inline src/ecs/resources/broadcaster.rs]: I don't think this call is necessary, dunno why it is there.

File: src/engine/app.rs [hunk 4]
```diff
@@ -64,10 +65,11 @@ impl Application {
         assets.add_loader::<gfx_types::Factory>(factory);
 
         let trans_sys = TransformSystem::new();
-        planner.add_system::<TransformSystem>(trans_sys, "transform_system", 0);
+
+        let dispatcher_builder = dispatcher_builder.add_barrier();
```
Comment (torkleyy) [inline src/engine/app.rs]: As discussed on Gitter, it would be preferable to give the responsibility of adding "default" systems to the user.

File: src/engine/app.rs [hunk 5]
```diff
@@ -221,51 +221,57 @@ impl Application {
...
-impl<T> ApplicationBuilder<T>
+impl<'a, T> ApplicationBuilder<'a, T>
     where T: State + 'static
 {
     /// Creates a new ApplicationBuilder with the given initial game state and
     /// display configuration.
-    pub fn new(initial_state: T, cfg: DisplayConfig) -> ApplicationBuilder<T> {
-        let pool = Arc::new(ThreadPool::new(num_cpus::get()));
+    pub fn new(initial_state: T, cfg: DisplayConfig) -> ApplicationBuilder<'a, T> {
+        let pool = Arc::new(ThreadPool::new( ::rayon::Configuration::new().num_threads(num_cpus::get())).unwrap());
```
Comment (torkleyy) [inline src/engine/app.rs]: Could you format the code in some way? Seems there is unnecessary whitespace. And please add an import statement inside the function.

Additionally, consider using `expect` over `unwrap`.

File: src/engine/app.rs [hunk 6]
```diff
@@ -221,51 +221,57 @@ impl Application {
...
 
     /// Adds a given system `pro`, assigns it the string identifier `name`,
     /// and marks it with the runtime priority `pri`.
-    pub fn with<S>(mut self, sys: S, name: &str, pri: Priority) -> ApplicationBuilder<T>
-        where S: System<()> + 'static
+    pub fn with<S>(mut self, sys: S, name: &str, dep: &[&str]) -> ApplicationBuilder<'a, T>
+        where for<'b> S: System<'b> + Send + 'a
     {
-        self.planner.add_system::<S>(sys, name, pri);
+        self.dispatcher_builder = self.dispatcher_builder.add(sys, name, dep);
```
Comment (torkleyy) [inline src/engine/app.rs]: Not sure if we should just wrap the functionality. Could we just let the user pass in the dispatcher with all systems already added?

File: src/ecs/resources/broadcaster.rs (hunk 3 above)
Comment (omni-viral) [inline src/ecs/resources/broadcaster.rs]: I think I should move it to the end of `Application::advance_frame`

File: src/engine/app.rs (hunk 6 above)
Comment (omni-viral) [inline src/engine/app.rs]: Same way we could just let the user to pass world with all `Component`s added.
But there is interesting feature in @ebkalderon`s fork. `SystemExt` which can be used to add all related components. We could use it in `ApplicationBuilder`'s wrapping functions

Comment (torkleyy) [inline src/engine/app.rs]: Why not integrate it into Specs? I didn't know about that, but it would be helpful, yeah.

Comment (torkleyy) [inline src/engine/app.rs]: Oh well, it seems it isn't 100% compatible with our existing builders...

Comment (omni-viral) [inline src/engine/app.rs]: I think there should be two possibilities for user
Use `ApplicationBuilder` and add `System`s and register `Component`s through it.
Create `Application` instance directly passing `Dispatcher` with all `System`s and `World` with all `Components` added.

File: examples/04_pong/main.rs [hunk 7]
```diff
@@ -3,8 +3,10 @@ extern crate amethyst;
 use amethyst::{Application, Event, State, Trans, VirtualKeyCode, WindowEvent};
 use amethyst::asset_manager::AssetManager;
 use amethyst::project::Config;
-use amethyst::ecs::{World, Join, VecStorage, Component, RunArg, System};
+use amethyst::ecs::{World, Join, VecStorage, Component, System, WriteStorage, Fetch, FetchMut};
```
Comment (torkleyy) [inline examples/04_pong/main.rs]: Needs reordering.

File: examples/04_pong/main.rs [hunk 8]
```diff
@@ -120,13 +115,13 @@ impl System<()> for PongSystem {
...
                     // If `W` is pressed and plank is in screen boundaries then move up
-                    if input.key_down(VirtualKeyCode::W) {
+                    if input.key_is_pressed(VirtualKeyCode::W) {
                         if plank.position + plank.dimensions[1] / 2. < 1. {
                             plank.position += plank.velocity * delta_time;
                         }
                     }
                     // If `S` is pressed and plank is in screen boundaries then move down
-                    if input.key_down(VirtualKeyCode::S) {
+                    if input.key_is_pressed(VirtualKeyCode::S) {
```
Comment (torkleyy) [inline examples/04_pong/main.rs]: I think we need a rebase here.

File: src/engine/app.rs [hunk 9]
```diff
@@ -221,51 +217,70 @@ impl Application {
...
         self
     }
 
-    /// Adds a given system `pro`, assigns it the string identifier `name`,
-    /// and marks it with the runtime priority `pri`.
-    pub fn with<S>(mut self, sys: S, name: &str, pri: Priority) -> ApplicationBuilder<T>
-        where S: System<()> + 'static
+    /// Adds a given thread-local system `sys`
+    /// All thread-local systems executed sequentially after all non-thread-local systems
```
Comment (torkleyy) [inline src/engine/app.rs]: Maybe add an "are" in here?

File: src/lib.rs [hunk 10]
```diff
@@ -84,7 +84,7 @@ extern crate genmesh;
 extern crate imagefmt;
 extern crate num_cpus;
 extern crate specs;
-extern crate threadpool;
+extern crate rayon;
```
Comment (torkleyy) [inline src/lib.rs]: Sorry, have to nitpick here again - please move the rayon line up by one.

File: src/ecs/systems/transform.rs [hunk 11]
```diff
@@ -3,7 +3,7 @@
 use cgmath::Matrix4;
 use fnv::{FnvHashMap as HashMap, FnvHashSet as HashSet};
 
-use ecs::{Join, Entity, RunArg, System};
+use ecs::{Entity, Join, WriteStorage, ReadStorage, Entities, System};
```
Comment (torkleyy) [inline src/ecs/systems/transform.rs]: Needs reordering.

File: examples/05_assets/main.rs [hunk 12]
```diff
@@ -214,6 +213,8 @@ fn main() {
     let path = format!("{}/examples/05_assets/resources/config.yml",
                        env!("CARGO_MANIFEST_DIR"));
     let cfg = DisplayConfig::load(path);
-    let mut game = Application::build(Example, cfg).done();
+    let mut game = Application::build(Example, cfg)
+        .with_thread_local::<TransformSystem>(TransformSystem::new())
```
Comment (torkleyy) [inline examples/05_assets/main.rs]: Are you sure `TransformSystem` is thread-local?

File: examples/04_pong/main.rs [hunk 13]
```diff
@@ -342,7 +335,8 @@ fn main() {
     let mut game = Application::build(Pong, cfg)
         .register::<Ball>()
         .register::<Plank>()
-        .with::<PongSystem>(PongSystem, "pong_system", 1)
+        .with::<PongSystem>(PongSystem, "pong_system", &[])
+        .with_thread_local::<TransformSystem>(TransformSystem::new())
```
Comment (torkleyy) [inline examples/04_pong/main.rs]: Same here.

File: examples/03_renderable/main.rs [hunk 14]
```diff
@@ -292,7 +285,8 @@ fn main() {
                        env!("CARGO_MANIFEST_DIR"));
     let cfg = DisplayConfig::load(path);
     let mut game = Application::build(Example, cfg)
-        .with::<ExampleSystem>(ExampleSystem, "example_system", 1)
+        .with::<ExampleSystem>(ExampleSystem, "example_system", &[])
+        .with_thread_local::<TransformSystem>(TransformSystem::new())
```
Comment (torkleyy) [inline examples/03_renderable/main.rs]: Same here.

File: examples/05_assets/main.rs (hunk 12 above)
Comment (omni-viral) [inline examples/05_assets/main.rs]: Well. After you mention that I think it is not :)

Comment (omni-viral) [inline examples/05_assets/main.rs]: It just require a barrier

**Comment by omni-viral** [comment]
@ebkalderon There might be conflicts. You indeed may rebase and reject [ebkalderon/amethyst#2](https://github.com/ebkalderon/amethyst/pull/2)

**Comment by ebkalderon** [comment]
@omni-viral Sweet, I reverted it successfully. I will rebase on top of this PR once it's ready for merge!

**Comment by torkleyy** [comment]
@ebkalderon I'd like to review once more, if that's okay?

**Comment by ebkalderon** [comment]
Of course, go right ahead.

**Comment by omni-viral** [comment]
Thank you all for your feedback! 
I'm glad I can help :)

File: src/engine/app.rs [hunk 15]
```diff
@@ -221,51 +217,70 @@ impl Application {
...
+        use rayon::Configuration;
+
+        let pool = Arc::new(ThreadPool::new(Configuration::new().num_threads(num_cpus::get())).expect("Failed to create rayon::ThreadPool"));
 
         ApplicationBuilder {
             config: cfg,
             initial_state: initial_state,
-            planner: Planner::from_pool(World::new(), pool),
+            dispatcher_builder: DispatcherBuilder::new().with_pool(pool),
+            world: World::new(),
```
Comment (torkleyy) [inline src/engine/app.rs]: Hold on, shouldn't we give the user the chance to access the `World` before starting the game? Not exactly sure about it.

Comment (Xaeroxe) [inline src/engine/app.rs]: Yeah I think so.  There is a space between when the Application is built and when we call `.run()` on it, so all it would take is a adding a `.world_mut()` function to the Application.

Comment (torkleyy) [inline src/engine/app.rs]: Or passing in `World` instead of creating it inside the `ApplicationBuilder`.

Comment (Xaeroxe) [inline src/engine/app.rs]: Either or.  Both are functionally the same I think.

**Comment by ebkalderon** [comment]
@torkleyy Heads up! It's possible that this pull request might break your ongoing asset management work on #244. Please make sure to rebase against this one if any breakage occurs.

**Comment by torkleyy** [comment]
@ebkalderon How is that? I don't even depend on `amethyst` crate so far, so I don't think I will have any problems rebasing.
